docs/modules/ROOT/pages/gsn.adoc

```diff
@@ -0,0 +1,77 @@
+
+You don't need to! The GSN is set up in such a way where it's in the relays' best interest to serve your requests, and there are measures in place to penalize them if they misbeheave. All of this happens automatically, so you can safely start using their services worry-free.
+
+== One contract to coordinate them all
+
+There are many meta-transaction implementations out there, but the GSN has a unique detail that makes it special. Sitting in the middle of the system, a smart contract is charged with keeping track of relays, handling relayed transactions, charging their recipients, and generally ensuring all parties stay honest. This contract is called RelayHub, and there is a _single_ instance of it in the whole network (you don't need to deploy your own!). Think of it as a piece of public infrastructure, for all Ethereum users to benefit from.
+
+One of RelayHub's tasks is relay-coordination: all relays will advertise their services on this contract, and your users will query it to find the relay that best suits their purposes. This is out of scope for this guide however, and is not something you need to worry about when writing a recipient contract. If you want to learn more about sending transactions via relays, head to our [GSNProvider guide]().
+
+== Receiving a relayed call
+
+The contract you will be inheriting from is GSNRecipient. It includes the IRelayRecipient interface, plus a number of useful utilities that handle the low-level details of the GSN, so you can focus on your dApp.
+
+Unlike regular contract function calls, each relayed call has an additional number of steps it must go through. Most of writing a recipient involves handling these, which are designed to provide enough flexibility to cater to all possible use cases. Basic recipients can however safely ignore most of them, while still being secure and sound.
+
+The OpenZeppelin Contracts provide a number of [tried-and-tested approaches]() for you to use out of the box, but you should still have a basic idea of what's going on under the hood.
+
+==== acceptRelayedCall
+
+First, RelayHub will ask your recipient contract if it wants to receive a relayed call. Recall that you will be charged for incurred gas costs by the relay, so you should only accept calls that you're willing to pay for!
+
+There are multiple ways to make this work, including:
+ * having a whitelist of trusted users
+ * only accepting calls to an onboarding function
+ * charging users in tokens (possibly issued by you)
+ * delegating the acceptance logic off-chain
+
+All relayed call requests can be rejected at no cost to the recipient.
+
+=== pre and postRelayedCall
+
+You may have noticed one of the suggested acceptance flows involves charging a user in tokens. You may, for example, setup a system where users spend a special token to pay for relayed calls, and then either gift these to new users, or sell them off-chain. The GSN will let you easily implement these designs.
```

Comment:
I'd move the example _after_ we have explained what pre and post are, or even remove this in favor of a link to the advanced use cases guide.

Reply:
The issue is that all pre and post do is allow for user charging, so we should definitely cover it beforehand.

I see two possible ways to tackle this:
 1. go over the 'user charging' schemes earlier in the guide (there's currently a short snippet in the 'incentives' section)
 2. make pre and post optional by having GSNRecipient include empty implementations, and only go over them in the advanced used cases guide

Reply:
> The issue is that all pre and post do is allow for user charging, so we should definitely cover it beforehand.

True, but then I would start this paragraph with exactly that, instead of a direct reference to token payments. Otherwise, a reader may understand that it can only be used with an associated token.

I'd reorder this as follows:
1- Start with "After a relayed call is accepted...", since it ties very well with the previous section
2- Explain that these functions are usually used to charge your users, and that an example could be paying with a native token
3- Then go into the details of pre and post
4- Then some sample code